feat: Implement an add command

pulsejet opened the pull request:
`add` takes a document and adds it with a random ID to the specified collection. `Set` in firestore takes an ID as well, so I renamed the method here to `add` to avoid confusion. `Set` would be very similar to this I should think.

Commit:
feat: Implement an add command

Commit:
test(client): Add unit test for Add

pulsejet commented:
@mookjp I'm not sure if this test is good enough, but I can't think of anything better.

mookjp commented on lib/client.go:
```diff
@@ -86,3 +86,13 @@ func (c *Client) Query(ctx context.Context, collection string, conditions []*Con
 }
+
+// Add a document into a collection
+func (c *Client) Add(ctx context.Context, collection string, document interface{}) (*firestore.DocumentRef, error) {
```
Could you use `Doc` type as a return type so that it can be parsed as JSON via `String()` func?

pulsejet replied:
Done

mookjp commented on lib/client_test.go:
```diff
@@ -335,6 +335,88 @@ func TestClient_Query(t *testing.T) {
 	}
 }
 
+func TestClient_Add(t *testing.T) {
```
Please use `Helper` for teardown to delete all data.
You can use `DeleteAll`.
It may have the effect to other test cases if you have made some changes in a test case.

pulsejet replied:
I had to create another `Helper` object to get this to work (the global one throws a `the client connection is closing` error). Is this okay?

mookjp replied:
Yes, please replace the helper to use it!

pulsejet replied:
Ooh got it now. Did that.

mookjp commented on cmd/add.go:
```diff
@@ -0,0 +1,68 @@
+		}
+
+		getCtx, _ := context.WithCancel(ctx)
+		res, err := client.Add(getCtx, collectionName, doc)
```
Please use `Doc` and just print out JSON as a result so that the user can see the formated result with `jq` command.
See `README` how it should be worked with it.

pulsejet replied:
Ah I see. Done.

mookjp commented on lib/client_test.go:
```diff
@@ -335,6 +335,88 @@ func TestClient_Query(t *testing.T) {
+			wantErr: false,
+		},
+		{
+			name: "returns an error when the document is nil",
```
Is there a possibility that the input is nil in this case? I don't think so.

pulsejet replied:
Since this function is taking an interface unmarshalled from JSON, it will be `nil` if the JSON string is invalid (or empty). I've added an explicit check for this in `cmd/add.go` though, so this might not be necessary.

mookjp replied:
> Since this function is taking an interface unmarshalled from JSON

I think it returns `error` and it should be handled?
Rather than adding the case like that, it seems better to confirm like:

* What should it do in the case if the root level is `Array`?
  * It is the correct JSON format but Firestore won't support it, I guess?

Your idea if not bad but it would be better to think about what kind of input will be put by the user as a parameter 😀

pulsejet replied:
> It is the correct JSON format but Firestore won't support it, I guess?

Right. Fixed this by changing it to a `map[string]interface{}` instead, so json unmarshalling will fail for `Array` (or any other invalid input).

> I think it returns error and it should be handled?

Didn't get this point. Should I check for `nil` and make a `New` error in `client.go`?

mookjp replied:
> Didn't get this point. Should I check for nil and make a New error in client.go?


No, I just meant that it would be better to have a possible case and it is also ok not to remove the case anyway.

mookjp commented on cmd/add.go:
```diff
@@ -0,0 +1,68 @@
+			os.Exit(1)
+		}
+
+		getCtx, _ := context.WithCancel(ctx)
```
`addCtx`?

pulsejet replied:
Sorry 😅

mookjp reviewed (changes requested):
Could you add documentation to `README` also? 🙏

Commit:
fix(add): change return type to *Doc

Commit:
fix(client): delete all after add test

Commit:
refactor(add): getCtx --> addCtx

Commit:
docs(readme): Update for add

Commit:
fix(add): Println --> Printf

Commit:
fix(add): Add check for invalid JSON

pulsejet commented:
@mookjp thanks for the review! Updated readme and fixed up everything (I think).

Commit:
refactor(test): replace helper variable

mookjp reviewed (commented):
@pulsejet Commented!

Commit:
fix(add): change document type to map[string]interface{}
Firestore doesn't allow plain arrays to be stored, so that is an
unhandled situation

Commit:
test(add): remove nil test
